selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/theia/TheiaBuildPluginTest.java

```diff
@@ -91,38 +74,174 @@ public void tearDown() throws Exception {
+  }
+
+  @Test(priority = 1)
+  public void hostedModeShouldWork() {
+    final String projectName = "hello-world";
+    final String hostedSearchSequence = ">hosted";
+    final String editedSourceFile = "hello-world-backend.ts";
+    final String expectedHelloWorldNotification = "Hello World!";
+    final String expectedAlohaWorldNotification = "Aloha World!";
+    final String suggestionForSelection = "Hosted Plugin: Start Instance";
+    final String helloWorldProposal = "Hello World";
+    final String helloWorldSearchSequence = ">Hello";
+    final String expectedInstanceRunningMessage = "Hosted instance is running at:";
+    final String parentWindow = seleniumWebDriver.getWindowHandle();
+    final String expectedPluginFolderMessage =
+        "Plugin folder is set to: file:///projects/hello-world";
+    final String expectedStartingServerMessage = "Starting hosted instance server ...";
+    final String editedSourceLine = "theia.window.showInformationMessage('Aloha World!');";
+    final String expectedTextAfterDeleting =
+        "context.subscriptions.push(theia.commands.registerCommand(informationMessageTestCommand, (...args: any[]) => {\n"
+            + " \n"
+            + "    }));";
+
+    // run hosted mode
+    theiaProjectTree.waitItem(projectName);
+    theiaProjectTree.clickOnItem(projectName);
+    theiaProjectTree.waitItemSelected(projectName);
+    WaitUtils.sleepQuietly(2);
+    theiaIde.pressKeyCombination(Keys.LEFT_CONTROL, Keys.LEFT_SHIFT, "p");
+    theiaProposalForm.waitForm();
+    WaitUtils.sleepQuietly(4);
+    theiaProposalForm.enterTextToSearchField(hostedSearchSequence);
+    WaitUtils.sleepQuietly(4);
+    theiaProposalForm.clickOnProposal(suggestionForSelection);
+    hostedPluginSelectPathForm.waitForm();
+    hostedPluginSelectPathForm.clickOnItem(projectName);
+    hostedPluginSelectPathForm.waitItemSelected(projectName);
+    hostedPluginSelectPathForm.clickOnOpenButton();
+    hostedPluginSelectPathForm.waitFormClosed();
+    waitNotificationEqualsTo(expectedPluginFolderMessage, parentWindow);
+    waitNotificationEqualsTo(expectedStartingServerMessage, parentWindow);
+    waitNotificationContains(expectedInstanceRunningMessage, parentWindow);
+
+    // check hosted mode availability
+    switchToNonParentWindow(parentWindow);
+    waitDevelopmentHostTitle();
+    theiaProjectTree.clickOnFilesTab();
+    theiaProjectTree.waitProjectAreaOpened();
+    theiaProjectTree.waitOpenWorkspaceButton();
+
+    // check plugin output
+    theiaIde.pressKeyCombination(Keys.LEFT_CONTROL, Keys.LEFT_SHIFT, "p");
+    theiaProposalForm.waitSearchField();
+    theiaProposalForm.enterTextToSearchField(helloWorldSearchSequence);
+    theiaProposalForm.clickOnProposal(helloWorldProposal);
+    theiaIde.waitNotificationEqualsTo(expectedHelloWorldNotification);
+    theiaIde.waitNotificationDisappearance(expectedHelloWorldNotification);
+
+    // check editing of the source code and applying it in hosted mode window
```

Comment:
It would make the test clear if you separate the steps which `applying it in hosted mode window`.